pkg/apis/stork/v1alpha1/resourcetransformation.go

```diff
@@ -0,0 +1,116 @@
+	// in case slice, entries will be appended.
+	// in case keypair, entries will be merged
+	ModifyPathValue ResourceTransformationOperationType = "modify"
+	// DeletePath from resource specification
+	DeleteResourcePath ResourceTransformationOperationType = "delete"
+	// JsonResourcePatch will path json in given resource spec
+	JsonResourcePatch ResourceTransformationOperationType = "jsonpatch"
+)
+
+// ResourceTransformationValueType is types of value supported on
+// path in resource specs
+type ResourceTransformationValueType string
+
+const (
+	// IntResourceType is to update integer value to specified resource path
+	IntResourceType ResourceTransformationValueType = "int"
+	// StringResourceType is to update string value to specified resource path
+	StringResourceType ResourceTransformationValueType = "string"
+	// BoolResourceType is to update boolean value to specified resource path
+	BoolResourceType ResourceTransformationValueType = "bool"
+	// SliceResourceType is to update slice value to specified resource path
+	SliceResourceType ResourceTransformationValueType = "slice"
+	// KeyPairResourceType is to update keypair value to specified resource path
+	KeyPairResourceType ResourceTransformationValueType = "keypair"
+)
+
+// ResourceTransformationStatsusType is status of resource transformation CR
+type ResourceTransformationStatusType string
+
+const (
+	// ResourceTransformationStatusInitial represents initial state of resource
+	// transformation CR
+	ResourceTransformationStatusInitial ResourceTransformationStatusType = "int"
+	// ResourceTransformationStatusInProgress represents dry run in progress state
+	// of resource transformation
+	ResourceTransformationStatusInProgress ResourceTransformationStatusType = "int"
+	// ResourceTransformationStatusReady represents ready state of resource
+	// transformation CR
+	ResourceTransformationStatusReady ResourceTransformationStatusType = "ready"
+	// ResourceTransformationStatusError represents dry-run failed state of resource
+	// transformation CR
+	ResourceTransformationStatusError ResourceTransformationStatusType = "error"
+)
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ResourceTransformation represents a ResourceTransformation CR object
+type ResourceTransformation struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec              ResourceTransformationSpec       `json:"specs"`
+	Status            ResourceTransformationStatusType `json:"status"`
+}
+
+// ResourceTransformationSpec is used to update k8s resources
+//before migration/restore
+type ResourceTransformationSpec struct {
+	Objects []TransformSpecs `json:"transformSpecs"`
+}
+
+// TransformSpecs specifies the patch to update selected resource
+// before migration/restore
+type TransformSpecs struct {
+	// Selectors label selector to filter out resource for
+	// patching
+	Selectors map[string]string `json:"selectors"`
```

Comment:
Should we take an object type here as well (GVK)? A complete application stack involving a PVC, Service, Deployment, ConfigMap can potentially have the same selectors, but we might want a transformation only on a Service object. And the type can take a wildcard `*` which would allow us to apply to all the resources if required.